agent/proxycfg/state.go

```diff
@@ -345,6 +344,16 @@ func (s *state) run() {
 		Proxy:           s.proxyCfg,
 		Datacenter:      s.source.Datacenter,
 	}
+}
+
+func (s *state) run() {
+	// Close the channel we return from Watch when we stop so consumers can stop
+	// watching and clean up their goroutines. It's important we do this here and
+	// not in Close since this routine sends on this chan and so might panic if it
+	// gets closed from another goroutine.
+	defer close(s.snapCh)
+
+	snap := s.initialConfigSnapshot()
 
 	switch s.kind {
```

Comment:
I think this switch statement needs to go into the `initialConfigSnapshot` method body. It's the rest of the setup.